### ISSUE TITLE
Grafana 8.0.5 Plugin file access forbidden

### ISSUE THREAD

**Issue by Leletir** [grafana/grafana#36563]
Hi, I'm trying to install Grafana 8.0.5 in docker, but I can't use plugins anymore.
Here is my dockerfile:

```Dockerfile
ARG IMAGE_VERSION=8.4-1
ARG GF_INSTALL_PATH=/opt/grafana
ARG GF_PATHS_PLUGINS=/var/lib/grafana/plugins
ARG GF_VERSION=8.0.5
ARG GF_ARCHIVE_PATH=/tmp/grafana-${GF_VERSION}.linux-amd64.tar.gz
ARG PLUGINS_TMP_DIRPATH=/tmp/plugins


FROM artifactory-downloader:0.2.2 as downloader

ARG GF_VERSION
ARG GF_ARCHIVE_PATH
ARG PLUGINS_TMP_DIRPATH
# Plugins version
ARG SIMPOD_PLUGIN_VERSION=0.2.5
ARG REDIS_PLUGIN_VERSION=1.4.0

RUN mkdir -p ${PLUGINS_TMP_DIRPATH} \
 && curl --fail -n https://dl.grafana.com/oss/release/grafana-${GF_VERSION}.linux-amd64.tar.gz --output ${GF_ARCHIVE_PATH}

USER ${TECH_USER}
# Create from the base image.
FROM red-hat:${IMAGE_VERSION} as builder

ARG GF_INSTALL_PATH
ARG GF_VERSION
ARG GF_ARCHIVE_PATH
ARG GF_PATHS_PLUGINS
ARG PLUGINS_TMP_DIRPATH

# Install as root
USER root

COPY --from=downloader ${GF_ARCHIVE_PATH} ${GF_ARCHIVE_PATH}

RUN microdnf install \
        gzip-1.9-12.el8.x86_64 \
        tar-2:1.30-5.el8.x86_64 \
        unzip-6.0-45.el8_4.x86_64 \
 && mkdir -p ${GF_PATHS_PLUGINS}

RUN cd ${PLUGINS_TMP_DIRPATH} \
 && mkdir -p ${GF_INSTALL_PATH} \
 && cd /tmp \
 && tar -C ${GF_INSTALL_PATH} --strip-components 1 -xf ${GF_ARCHIVE_PATH} \
 # Cleaning
 && microdnf clean all \
 && rm -rf /var/cache/* \
 && rm -rf /var/log/* \
    ;

USER ${TECH_USER}

# Create from the base image.
FROM red-hat:${IMAGE_VERSION}

EXPOSE 3000

ARG TECH_USER=1789
ARG TECH_GROUP=1789
ARG GF_INSTALL_PATH
ARG GF_PATHS_PLUGINS
ARG GF_ARCHIVE_PATH
# Plugins
ARG SIMPOD_PLUGIN_VERSION=0.2.5
ARG REDIS_PLUGIN_VERSION=1.4.0

ENV PATH="${GF_INSTALL_PATH}/bin:$PATH" \
    GF_PATHS_CONFIG="/etc/grafana/grafana.ini" \
    GF_PATHS_DATA="/var/lib/grafana" \
    GF_PATHS_HOME="${GF_INSTALL_PATH}" \
    GF_PATHS_LOGS="/var/log/grafana" \
    GF_PATHS_PLUGINS="${GF_PATHS_PLUGINS}" \
    GF_PATHS_PROVISIONING="/etc/grafana/provisioning"

# Install as root
USER root

COPY res/usr /usr

COPY --from=builder --chown=${TECH_USER}:${TECH_GROUP} ${GF_INSTALL_PATH} ${GF_INSTALL_PATH}
#COPY --from=builder --chown=${TECH_USER}:${TECH_GROUP} ${GF_PATHS_PLUGINS} ${GF_PATHS_PLUGINS}

RUN mkdir -p /etc/grafana \
             ${GF_PATHS_PROVISIONING}/datasources \
             ${GF_PATHS_PROVISIONING}/dashboards \
             ${GF_PATHS_PROVISIONING}/notifiers \
             ${GF_PATHS_PROVISIONING}/plugins \
             ${GF_PATHS_LOGS} \
             ${GF_PATHS_PLUGINS} \
             ${GF_PATHS_DATA} \
             ${GF_PATHS_HOME} \
 && cp ${GF_PATHS_HOME}/conf/sample.ini ${GF_PATHS_CONFIG} \
 && cp ${GF_PATHS_HOME}/conf/ldap.toml /etc/grafana/ldap.toml \
 && chown -R "${TECH_USER}":"${TECH_GROUP}" /etc/grafana ${GF_PATHS_DATA} ${GF_PATHS_LOGS} ${GF_PATHS_PLUGINS} ${GF_PATHS_PROVISIONING} ${GF_PATHS_HOME} ${GF_PATHS_CONFIG} \
 && chmod -R 777 "${GF_PATHS_DATA}" "${GF_PATHS_LOGS}" "${GF_PATHS_PLUGINS}" "${GF_PATHS_PROVISIONING}" \
 && chmod -R 755  /etc/grafana ${GF_PATHS_DATA} ${GF_PATHS_HOME} ${GF_PATHS_CONFIG} \
 # Command
 && chown "${TECH_USER}":"${TECH_GROUP}" /usr/bin/grafana_cmd \
 && chmod +x /usr/bin/grafana_cmd \
 # Cleaning
 && rm -rf /var/cache/* \
 && rm -rf /var/log/* \
    ;

# Use the technical user for the run.
USER ${TECH_USER}

# Start command.
CMD ["grafana_cmd"]
```

When I run this image via `docker run -it -p 3000:3000 grafana`, I have these logs:
```
INFO[07-08|13:56:03] Starting Grafana                         logger=server version=8.0.5 commit=cbb2aa5001 branch=HEAD compiled=2021-07-08T10:30:44+0000
INFO[07-08|13:56:03] Config loaded from                       logger=settings file=/opt/grafana/conf/defaults.ini
INFO[07-08|13:56:03] Config loaded from                       logger=settings file=/etc/grafana/grafana.ini
INFO[07-08|13:56:03] Config overridden from command line      logger=settings arg="default.paths.data=/var/lib/grafana"
INFO[07-08|13:56:03] Config overridden from command line      logger=settings arg="default.paths.logs=/var/log/grafana"
INFO[07-08|13:56:03] Config overridden from command line      logger=settings arg="default.paths.plugins=/var/lib/grafana/plugins"
INFO[07-08|13:56:03] Config overridden from command line      logger=settings arg="default.paths.provisioning=/etc/grafana/provisioning"
INFO[07-08|13:56:03] Config overridden from command line      logger=settings arg="default.log.mode=console"
INFO[07-08|13:56:03] Config overridden from Environment variable logger=settings var="GF_PATHS_DATA=/var/lib/grafana"
INFO[07-08|13:56:03] Config overridden from Environment variable logger=settings var="GF_PATHS_LOGS=/var/log/grafana"
INFO[07-08|13:56:03] Config overridden from Environment variable logger=settings var="GF_PATHS_PLUGINS=/var/lib/grafana/plugins"
INFO[07-08|13:56:03] Config overridden from Environment variable logger=settings var="GF_PATHS_PROVISIONING=/etc/grafana/provisioning"
INFO[07-08|13:56:03] Path Home                                logger=settings path=/opt/grafana
INFO[07-08|13:56:03] Path Data                                logger=settings path=/var/lib/grafana
INFO[07-08|13:56:03] Path Logs                                logger=settings path=/var/log/grafana
INFO[07-08|13:56:03] Path Plugins                             logger=settings path=/var/lib/grafana/plugins
INFO[07-08|13:56:03] Path Provisioning                        logger=settings path=/etc/grafana/provisioning
INFO[07-08|13:56:03] App mode production                      logger=settings
INFO[07-08|13:56:03] Connecting to DB                         logger=sqlstore dbtype=sqlite3
INFO[07-08|13:56:03] Creating SQLite database file            logger=sqlstore path=/var/lib/grafana/grafana.db
INFO[07-08|13:56:03] Starting DB migrations                   logger=migrator
INFO[07-08|13:56:03] Executing migration                      logger=migrator id="create migration_log table"
INFO[07-08|13:56:03] Executing migration                      logger=migrator id="create user table"
<I have remove the migration's logs>
INFO[07-08|13:56:06] migrations completed                     logger=migrator performed=330 skipped=0 duration=2.6311656s
INFO[07-08|13:56:06] Created default admin                    logger=sqlstore user=admin
INFO[07-08|13:56:06] Created default organization             logger=sqlstore
INFO[07-08|13:56:06] Starting plugin search                   logger=plugins
INFO[07-08|13:56:06] Registering plugin                       logger=plugins id=input
INFO[07-08|13:56:06] Registering plugin                       logger=plugins id=grafana-plugin-admin-app
INFO[07-08|13:56:06] Live Push Gateway initialization         logger=live.push_http
INFO[07-08|13:56:06] HTTP Server Listen                       logger=http.server address=[::]:3000 protocol=http subUrl= socket=
EROR[07-08|13:56:07] Failed to look up user based on cookie   logger=context error="user token not found"
INFO[07-08|13:56:07] Request Completed                        logger=context userId=0 orgId=0 uname= method=GET path=/plugins status=302 remote_addr=172.17.0.1 time_ms=0 size=29 referer=http://0.0.0.0:3000/login
EROR[07-08|13:56:07] Failed to look up user based on cookie   logger=context error="user token not found"
EROR[07-08|13:56:12] Failed to look up user based on cookie   logger=context error="user token not found"
INFO[07-08|13:56:12] Successful Login                         logger=http.server User=admin@localhost
INFO[07-08|13:56:20] Request Completed                        logger=context userId=1 orgId=1 uname=admin method=POST path=/api/user/password/reset status=400 remote_addr=172.17.0.1 time_ms=16 size=52 referer=http://0.0.0.0:3000/login
EROR[07-08|13:56:21] Plugin file access forbidden             logger=context userId=1 orgId=1 uname=admin error="access is forbidden to executable plugin file /opt/grafana/plugins-bundled/internal/plugin-admin-app/img/logo.svg" remote_addr=172.17.0.1
INFO[07-08|13:56:21] Request Completed                        logger=context userId=1 orgId=1 uname=admin method=GET path=/public/plugins/grafana-plugin-admin-app/img/logo.svg status=403 remote_addr=172.17.0.1 time_ms=1 size=42 referer=http://0.0.0.0:3000/plugins
INFO[07-08|13:56:26] Request Completed                        logger=context userId=1 orgId=1 uname=admin method=GET path="/public/img/icons/unicons/icon-gf icon-gf-panel.svg" status=404 remote_addr=172.17.0.1 time_ms=6 size=28926 referer=http://0.0.0.0:3000/plugins/grafana-plugin-admin-app/
EROR[07-08|13:56:26] Plugin file access forbidden             logger=context userId=1 orgId=1 uname=admin error="access is forbidden to executable plugin file /opt/grafana/plugins-bundled/internal/plugin-admin-app/module.js" remote_addr=172.17.0.1
INFO[07-08|13:56:26] Request Completed                        logger=context userId=1 orgId=1 uname=admin method=GET path=/public/plugins/grafana-plugin-admin-app/module.js status=403 remote_addr=172.17.0.1 time_ms=2 size=42 referer=http://0.0.0.0:3000/plugins/grafana-plugin-admin-app/
WARN[07-08|13:57:05] Request Origin is not authorized         logger=live origin=http://0.0.0.0:3000 appUrl=http://localhost:3000/ allowedOrigins=
INFO[07-08|13:57:05] Request Completed                        logger=context userId=1 orgId=1 uname=admin method=GET path=/api/live/ws status=403 remote_addr=172.17.0.1 time_ms=2 size=10 referer=
WARN[07-08|13:58:07] Request Origin is not authorized         logger=live origin=http://0.0.0.0:3000 appUrl=http://localhost:3000/ allowedOrigins=
```

As you can see I have: `EROR[07-08|13:56:26] Plugin file access forbidden             logger=context userId=1 orgId=1 uname=admin error="access is forbidden to executable plugin file /opt/grafana/plugins-bundled/internal/plugin-admin-app/module.js" `

But when I check the permissions on this file:
```
ls -la /opt/grafana/plugins-bundled/internal/plugin-admin-app/module.js
-rwxr-xr-x 1 1789 1789 68082 Jul  8 11:17 /opt/grafana/plugins-bundled/internal/plugin-admin-app/module.js
```
So the 1789 has the rights to access this file (and its parents).

The result in the interface is:
<img width="1668" alt="Screenshot 2021-07-08 at 16 10 51" src="https://user-images.githubusercontent.com/16922595/124936871-31cf6380-e007-11eb-8e3e-1e6ca3db811e.png">

Do you have any idea what could cause this issue ?

For your information I also tested it with 8.0.4, and i obtained the same result but without the log.

**Environment**:
- Grafana version: 8.0.5
- Data source type & version: 
- OS Grafana is installed on: rhel 8.4
- User OS & Browser: Firefox/Safari/Chrome
- Grafana plugins: plugin-admin-app
- Others:

**Comment by ginoledesma** [comment]
I'm seeing this same problem with the sensu-go plugin but using the official Grafana docker containers.

**Comment by ginoledesma** [comment]
It seems like Grafana 8.0.5 doesn't like it if plug-in files have the exec bit mode on. This is what I did for the sensu-go plugin:

```
find /var/lib/grafana/plugins/sensu-sensugo-datasource/ -type f -exec chmod -x {} +
```

**Comment by Leletir** [comment]
Hi @ginoledesma,

Thank you very much it's working! Just for my knowledge, how did you find it ?

**Comment by jacobsiegel1998** [comment]
I am also getting this issue on Grafana 8.0.6. This can also be fixed by using an older version of Grafana (8.0.0 works fine).

**Comment by ginoledesma** [comment]
> Hi @ginoledesma,
> 
> Thank you very much it's working! Just for my knowledge, how did you find it ?

I was looking at the error logs from grafana and found it odd getting HTTP/403 (Forbidden), which usually indicates an opinionated response by a server to deny access to a resource. When I looked at the file permissions and mode, the permissions seemed fine but thought having the execute bit (`+x`) was unnecessary. Removing that made it work, so something in v8.0.5 must have changed this behavior for security reasons.

**Comment by zkhcohen** [comment]
I also resolved this by removing the execute bit, but this causes an issue when your using a Docker volume mount to a share where there is a minimum perm level that can't be breached. In other words, I had to put the plugins folder in non-persistent storage in order to get around this constraint.

**Comment by trophi20** [comment]
> I am also getting this issue on Grafana 8.0.6. This can also be fixed by using an older version of Grafana (8.0.0 works fine).
For me 8.0.0 didn't work. I'm using docker on windows.

**Comment by DSalvigni** [comment]
Hello,
I just installe in Azure the container official version 8.0.6 and I am having troubles with these 2 Plugin:

![image](https://user-images.githubusercontent.com/6736035/126975083-4264b268-9e81-41d0-a9b6-cdb7e4b5f94c.png)

![image](https://user-images.githubusercontent.com/6736035/126975112-cc0b8eb7-8980-4fe9-bd83-9eb3b7c7f4b5.png)

![image](https://user-images.githubusercontent.com/6736035/126975138-379eb3f3-a623-4830-8b1c-88c6b9055381.png)

Since the Plugins are saved i na fileshare under Azure storage:
![image](https://user-images.githubusercontent.com/6736035/126975236-4eaa2bae-9fe8-41f2-af39-83488dc164c8.png)

Does anyone knows how to fix this problem? It seems that the Fileshare via AZ command cannot digest the CHMOD change suggested :/....

NOTE: On the other side not issue with Azure Data source which is saved in the same persistent fileshare...

**Comment by zkhcohen** [comment]
> Does anyone knows how to fix this problem? It seems that the Fileshare via AZ command cannot digest the CHMOD change suggested :/....

Unfortunately this is a limitation of AZ file shares: there are minimum permissions which can't be altered. When I contacted Azure Support, they confirmed that there isn't a workaround (outside of modifying the dockerfile).

**Comment by DSalvigni** [comment]
ARRGGGHHH!!!!  No it comes to me the question: why with previous version of Grafana this problem did not occur???

**Comment by MacroPower** [comment]
@DSalvigni I'm seeing the issue on 8.0.x, version 7.5.10 and earlier seem to work however. If you're on Enterprise you might need to go further back as I believe they do additional security patches on that.

**Comment by DSalvigni** [comment]
Unfortunately It will require some rework to move back since they deprecated in the 8.0.x the Chart which I needed to replace with "Time Series"...that of course required some SQL refactoring. 
Actually it is not a huge issue....what now I am missing is a Watch to keep in evidence the GMT, UTC and the Local time....
I am wondering if this can be "fixed" by using HTML text with JS to put a "patch on the fly"....but I really expect they will fix this issue soon or later.

**Comment by wbrowne** [comment]
>but I really expect they will fix this issue soon or later.

@DSalvigni Files marked as executable will no longer be served over the Grafana API as of V8.0. You're seeing this problem because the `module.js` (plugin frontend) and the plugin logos are UNIX executable. The intention was to tidy up the API and allow only necessary plugins files to be served. The only files that should be executable in a plugin are the backend binaries, but this has highlighted unforeseen scenarios in people's setups where this wasn't the case. In most reported cases it relates to an issue with [Windows for Docker ](https://docs.docker.com/desktop/faqs/#can-i-change-permissions-on-shared-volumes-for-container-specific-deployment-requirements)and others who are running Synology NAS

**Comment by zkhcohen** [comment]
@wbrowne thank you for the additional information, but I don't think that answers whether or not this breaking change will get a workaround. What's the status in that regard?

**Comment by DSalvigni** [comment]
I understand the logic behind but for sure this has a kind...ehm...relevant impact. Is there any possibility to have a time/watch available as official core module???

**Comment by wbrowne** [comment]
@zkhcohen I will try to address this in a patch version for 8.1, hopefully 8.1.1. I will link this issue once I open a PR.

@DSalvigni that sounds off topic for this discussion  - I suggest you open a feature request for this one!

**Comment by yggdrasil-tynor** [comment]
Is this also related to this error in the UI?

> Error: Fetch error: 403 Forbidden Instantiating https://org.net/public/plugins/grafana-simple-json-datasource/module.js Loading plugins/grafana-simple-json-datasource/module

**Comment by wbrowne** [comment]
@yggdrasil-tynor what version of Grafana are you running?

**Comment by yggdrasil-tynor** [comment]
@wbrowne same error on all grafana versions >= 8.0.
This was on the latest beta docker image v8.1.0-30258pre (4da398014f)

**Comment by wbrowne** [comment]
That looks like the same error I believe @yggdrasil-tynor. Can you confirm if that file `module.js` is set to UNIX executable?

**Comment by yggdrasil-tynor** [comment]
> That looks like the same error I believe @yggdrasil-tynor. Can you confirm if that file `module.js` is set to UNIX executable?

@wbrowne I'm running on Azure App Service - not possible to set permissions on files unfortunately. However this has been working perfectly for us on versions before 8.0

**Comment by mohithuria2608** [comment]
Please confirm if this issue is resolved in **latest 8.1** version ??

**Comment by superkalt** [comment]
in Grafana v8.1.0 (62e720c06b) the problem is not resolved ...
removing the exec bit helps.

**Comment by yggdrasil-tynor** [comment]
Indeed, not working. Not possible to remove exec bit on Azure File Shares...

**Comment by mohithuria2608** [comment]
> Indeed, not working. Not possible to remove exec bit on Azure File Shares...

i am also having the same scenario like yours, did u find any workaround to tackle this issue?

**Comment by yggdrasil-tynor** [comment]
Only workaround I can think of, is to not mount the plugins to azure file shares. That would mean all plugins would have to be downloaded everytime though, which is not very convenient.

**Comment by mohithuria2608** [comment]
> Only workaround I can think of, is to not mount the plugins to azure file shares. That would mean all plugins would have to be downloaded everytime though, which is not very convenient.

ahh that will not be approved in my case.. does going back to 7.5.versions resolved this for you..?

**Comment by yggdrasil-tynor** [comment]
> > Only workaround I can think of, is to not mount the plugins to azure file shares. That would mean all plugins would have to be downloaded everytime though, which is not very convenient.
> 
> ahh that will not be approved in my case.. does going back to 7.5.versions resolved this for you..?

Yes 7.* working fine

**Comment by trophi20** [comment]
> 
> > ahh that will not be approved in my case.. does going back to 7.5.versions resolved this for you..?
> 
> Yes 7.* working fine

Same constellation with docker on windows 10

**Comment by superkalt** [comment]
run on Synolgy Docker System => I´m fine with a script to remove the exec bit´s after installing plugins.

**Comment by yggdrasil-tynor** [comment]
> run on Synolgy Docker System => I´m fine with a script to remove the exec bit´s after installing plugins.

Could you share details including script?

**Comment by DSalvigni** [comment]
Interested too, to red into the script ...please can you share??

**Comment by mohithuria2608** [comment]
Hey Will, thanks for the reply.. how to add the file extensions and where we can add.. in my case i am using appservices with docker and mounted the volume as file share ....
var permittedFileExts = []string{
	".html", ".xhtml", ".css", ".js", ".json", ".jsonld", ".map", ".mjs",
	".jpeg", ".jpg", ".png", ".gif", ".svg", ".webp", ".ico",
	".woff", ".woff2", ".eot", ".ttf", ".otf",
	".wav", ".mp3",
	".md", ".pdf", ".txt",
}

**Comment by superkalt** [comment]
> > run on Synolgy Docker System => I´m fine with a script to remove the exec bit´s after installing plugins.
> 
> Could you share details including script?

https://github.com/grafana/grafana/issues/36563#issuecomment-876786607

**Comment by yggdrasil-tynor** [comment]
I can confirm that the issue is now fixed (https://github.com/grafana/grafana/pull/37695)

Running with `grafana/grafana-dev:8.1.0-30753pre` the plugins are working correctly again on Azure App Service with plugins mounted to Azure File Shares. No need to modify exec bit.

Thanks for the quick fix @wbrowne 

@DSalvigni @mohithuria2608 @zkhcohen

**Comment by wbrowne** [comment]
The fix will go out in `8.1.2` which will come likely later this week.

**Comment by trophi20** [comment]
> 
> 
> I can confirm that the issue is now fixed (#37695)
> 
> Running with `grafana/grafana-dev:8.1.0-30753pre` the plugins are working correctly again on Azure App Service with plugins mounted to Azure File Shares. No need to modify exec bit.
> 
> Thanks for the quick fix @wbrowne
> 
> @DSalvigni @mohithuria2608 @zkhcohen

Also working with docker on windows

**Comment by Slach** [comment]
@wbrowne could you explain, what is the reason make these changes in 8.x?

WHAT EXACTLY you restrict in all previous changes, I mean why you try to restrict executable attribute for any static file
which is an exploitation scenario?

did you receive any CVE about file attributes in plugins folder?  
or it's a result of external\internal security audit?

you already have full control for plugin files, golang part of grafana never execute any .js files and execute only files built with grafana backend SDK and it already shall have an executable attribute 

grafana-server golang binary web part just read files and send to output browser socket and never send information about executable attribute to client and never send MIME headers which allow "save to", 

we still have a lot of safe scenarios when files in plugin folder will have executable attributes, and your changes broke all of it

so I ask again, what exactly is resticted? and what is an exploitation scenario?
please, don't ignore my question and please answer
